src/solvers/dg_curved/dg_1d.jl

```diff
@@ -32,16 +35,64 @@ function calc_interface_flux!(u::AbstractArray{<:Any,3}, mesh::CurvedMesh{1},
   @threaded for element in eachelement(dg, cache)
     left_element = cache.elements.left_neighbors[1, element]
 
-    u_ll = get_node_vars(u, equations, dg, nnodes(dg), left_element)
-    u_rr = get_node_vars(u, equations, dg, 1,          element)
+    if left_element > 0 # left_element = -1 at bounaries
+      u_ll = get_node_vars(u, equations, dg, nnodes(dg), left_element)
+      u_rr = get_node_vars(u, equations, dg, 1,          element)
 
-    flux = surface_flux(u_ll, u_rr, 1, equations)
+      flux = surface_flux(u_ll, u_rr, 1, equations)
 
-    for v in eachvariable(equations)
-      cache.elements.surface_flux_values[v, 2, left_element] = flux[v]
-      cache.elements.surface_flux_values[v, 1, element] = flux[v]
+      for v in eachvariable(equations)
+        cache.elements.surface_flux_values[v, 2, left_element] = flux[v]
+        cache.elements.surface_flux_values[v, 1, element] = flux[v]
+      end
     end
   end
 
   return nothing
 end
+
+
+# TODO: Taal dimension agnostic
+function calc_boundary_flux!(cache, u, t, boundary_condition::BoundaryConditionPeriodic,
+                             equations::AbstractEquations{1}, mesh::CurvedMesh{1}, dg::DG)
+  @assert mesh.periodicity
+end
+
+
+# calc_boundary_flux! with boundary_condition::BoundaryConditionPeriodic is in dg.jl
+function calc_boundary_flux!(cache, u, t, boundary_conditions::Union{NamedTuple,Tuple},
+                             equations::AbstractEquations{1}, mesh::CurvedMesh{1}, dg::DG)
+  @unpack surface_flux = dg
+  @unpack surface_flux_values, node_coordinates = cache.elements
+
+  orientation = 1
+
+  # Negative x-direction
+  direction = 1
+
+  for j in eachnode(dg)
+    u_rr = get_node_vars(u, equations, dg, 1, 1)
+    x = get_node_coords(node_coordinates, equations, dg, 1, 1)
+
+    flux = boundary_conditions[direction](u_rr, orientation, direction, x, t, surface_flux, equations)
+
+    for v in eachvariable(equations)
+      surface_flux_values[v, direction, 1] = flux[v]
+    end
+  end
+
+  # Positive x-direction
+  direction = 2
+
+  for j in eachnode(dg)
+    u_rr = get_node_vars(u, equations, dg, nnodes(dg), nelements(dg, cache))
+    x = get_node_coords(node_coordinates, equations, dg, nnodes(dg), nelements(dg, cache))
+
+    flux = boundary_conditions[direction](u_rr, orientation, direction, x, t, surface_flux, equations)
+
+    # Copy flux to left and right element storage
+    for v in eachvariable(equations)
+      surface_flux_values[v, direction, nelements(dg, cache)] = flux[v]
+    end
+  end
+end
```

Comment:
Did you run some benchmarks to check the performance of this? Right now, I'm not sure whether the compiler is clever enough to figure out `boundary_conditions[direction]` - it should, but I would like to verify this. In particular, could you please check the performance in an example using at least two different functions/types in `boundary_conditions`? (It doesn't need to be in 1D, could be 2D as well - I just noticed it here at first.)

Reply:
I used
```julia
boundary_conditions = (boundary_condition_gauss, Trixi.boundary_condition_linear_y, Trixi.boundary_condition_linear_x_y, boundary_condition_gauss)
```
in `elixir_advection_amr_nonperiodic.jl` without AMR and in the corresponding curved elixir.
These are my results:
```
TreeMesh
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     4.229 μs (0.00% GC)
  median time:      4.271 μs (0.00% GC)
  mean time:        4.292 μs (0.00% GC)
  maximum time:     6.571 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     7

Old implementation without calc_boundary_flux_by_direction!
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     4.329 μs (0.00% GC)
  median time:      4.414 μs (0.00% GC)
  mean time:        4.432 μs (0.00% GC)
  maximum time:     7.429 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     7

New implementation with calc_boundary_flux_by_direction!
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     4.314 μs (0.00% GC)
  median time:      4.400 μs (0.00% GC)
  mean time:        4.418 μs (0.00% GC)
  maximum time:     7.843 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     7
```

Reply:
Thanks :+1: